### PR TITLE
LA-18: convert lectures to tsx

### LINKED ISSUE #17: Remove unused lectures
List:
- how-developers-work
- modern-java
- modern-php-features-and-ecosystem
- php-7-features-and-ecosystem

### LINKED ISSUE #18: Covert Lectures to tsx
Gradually  replace mdx with tsx

### PR DESCRIPTION
**Pull request by kulak91**

### CODE REVIEW

File: src/components/common/code/code.tsx [hunk 1]
```diff
@@ -0,0 +1,22 @@
+import * as React from 'react';
+import { Prism as SyntaxHighlighter } from 'react-syntax-highlighter';
+import { prism } from 'react-syntax-highlighter/dist/esm/styles/prism';
+import { CodeLanguage } from '~/enums/code-language/code-language';
+import { Copy } from '../copy/copy';
+import * as styles from './code.module.scss';
+
+type Props = {
+  language?: (typeof CodeLanguage)[keyof typeof CodeLanguage];
```
Comment (liza-veis) [inline src/components/common/code/code.tsx]:
```suggestion
  language?: CodeLanguage;
```

File: .eslintrc.yml [hunk 2]
```diff
@@ -15,13 +15,12 @@ extends:
 
 overrides:
   - files:
-    - gatsby-config.ts
-    - index.d.ts
-    - src/pages/**
-    - src/templates/**
+      - gatsby-config.ts
+      - index.d.ts
+      - src/pages/**
```
Comment (liza-veis) [inline .eslintrc.yml]: for these three lines indentation was correct

File: .lintstagedrc.yml [hunk 3]
```diff
@@ -1,3 +1,2 @@
 '*.{ts,tsx,js,jsx,json,css,scss}': prettier --write
-'*': npm run lint:editorconfig && npm run lint:fs
-'*.{ts, tsx}': npm run lint:js
+'*': npm run lint
```
Comment (liza-veis) [inline .lintstagedrc.yml]: we don't want to run lint:js, lint:type, lint:css for each files,only for needed ones, also this change is not related to this pr and should be discussed separately

File: package.json [hunk 4]
```diff
@@ -65,7 +63,7 @@
     "lint:js": "npx eslint . --resolve-plugins-relative-to .",
     "lint:fs": "ls-lint",
     "lint:editorconfig": "editorconfig-checker",
-    "lint": "npm run lint:js && npm run lint:type && npm run lint:css"
+    "lint": "npm run lint:js && npm run lint:type && npm run lint:css && npm run lint:fs"
```
Comment (liza-veis) [inline package.json]: add lint:editorconfig too

File: src/components/common/bubble/bubble.tsx [hunk 5]
```diff
@@ -0,0 +1,16 @@
+import * as React from 'react';
+import * as styles from './bubble.module.scss';
+import classNames from 'classnames';
+
+type Props = {
+  size?: 'small' | 'regular';
+};
+const Bubble: React.FC<React.PropsWithChildren<Props>> = ({
```
Comment (liza-veis) [inline src/components/common/bubble/bubble.tsx]: Define children in props
```suggestion
const Bubble: React.FC<Props> = ({
```

File: src/components/common/code/code.tsx [hunk 6]
```diff
@@ -0,0 +1,22 @@
+import * as React from 'react';
+import { Prism as SyntaxHighlighter } from 'react-syntax-highlighter';
+import { prism } from 'react-syntax-highlighter/dist/esm/styles/prism';
+import { CodeLanguage } from '~/enums/code-language/code-language';
+import { Copy } from '../copy/copy';
```
Comment (liza-veis) [inline src/components/common/code/code.tsx]:
```suggestion
import { Copy } from '~/components/common/common';
```

File: src/components/common/code/code.tsx [hunk 7]
```diff
@@ -0,0 +1,22 @@
+import * as React from 'react';
+import { Prism as SyntaxHighlighter } from 'react-syntax-highlighter';
+import { prism } from 'react-syntax-highlighter/dist/esm/styles/prism';
```
Comment (liza-veis) [inline src/components/common/code/code.tsx]:
```suggestion
import { prism as syntaxHighlighterStyles  } from 'react-syntax-highlighter/dist/esm/styles/prism';

File: src/components/common/code/readme.md
Comment (liza-veis) [inline src/components/common/code/readme.md:1]: Remove this file

File: src/components/common/copy/copy.tsx [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+type Props = {
+  text: string;
+};
+
+const Copy: React.FC<Props> = ({ text }) => {
+  const [copied, setCopied] = useState<boolean>(false);
+  const handleCopy = (): void => {
+    copy(text);
+    setCopied(true);
+    setTimeout(() => setCopied(false), 1500);
```
Comment (liza-veis) [inline src/components/common/copy/copy.tsx:18]:
```suggestion
```

File: src/components/common/intro-blue-bubble/intro-blue-bubble.tsx [hunk 9]
```diff
@@ -1,5 +1,5 @@
 import React from 'react';
-import { Avatar } from '~/components';
+import { Avatar } from '../avatar/avatar';
```
Comment (liza-veis) [inline src/components/common/intro-blue-bubble/intro-blue-bubble.tsx]:
```suggestion
import { Avatar } from '~/components/common';
```

File: src/components/common/level/level.tsx [hunk 10]
```diff
@@ -1,28 +1,15 @@
...
-  objectives: string;
-  labels?: Labels;
-};
-
 const Level: React.FC<Props> = ({
   number = '',
   name = '',
   difficulty = '',
   objectives = '',
+  id = '',
```
Comment (liza-veis) [inline src/components/common/level/level.tsx:12]: what do we need id for?

File: src/components/common/intro-blue-bubble/intro-blue-bubble.tsx (hunk 9 above)
Comment (liza-veis) [inline src/components/common/intro-blue-bubble/intro-blue-bubble.tsx]:
```suggestion
import { Avatar } from '~/components/common';
```

File: src/components/common/panels/panels.module.scss [hunk 11]
```diff
@@ -0,0 +1,57 @@
+@mixin generate-orders($max-index) {
+  @for $i from 0 to $max-index {
+    [data-columns='#{$i}'] {
+      grid-template-columns: repeat($i, minmax(0, 1fr));
+    }
+  }
+}
```
Comment (liza-veis) [inline src/components/common/panels/panels.module.scss]: make fixed columns

File: src/components/common/panels/panels.tsx [hunk 12]
```diff
@@ -0,0 +1,39 @@
...
+  bodySize?: 'small' | 'regular';
+};
+
+const Panels: React.FC<Props> = ({
+  panels,
+  columns = 2,
+  bodySize = 'regular',
+}) => {
+  return (
+    <div className={styles.wrapper} data-columns={columns}>
```
Comment (liza-veis) [inline src/components/common/panels/panels.tsx:21]:
```suggestion
    <div className={styles.wrapper}>
```

File: src/components/common/panels/panels.tsx [hunk 13]
```diff
@@ -0,0 +1,39 @@
...
+      {panels &&
+        panels.map(({ body, head }, index) => (
```
Comment (liza-veis) [inline src/components/common/panels/panels.tsx]:
```suggestion
      {panels.map(({ body, head }, index) => (
```

File: src/components/common/panels/panels.tsx [hunk 14]
```diff
@@ -0,0 +1,39 @@
...
+type Props = {
+  panels: Panel[];
+  columns?: number;
+  bodySize?: 'small' | 'regular';
+};
+
+const Panels: React.FC<Props> = ({
+  panels,
+  columns = 2,
+  bodySize = 'regular',
```
Comment (liza-veis) [inline src/components/common/panels/panels.tsx:18]: do we really need bodySize? looks like we can get rid of it

File: src/components/lectures/how-to-work-with-git/ua/index.tsx [hunk 15]
```diff
@@ -0,0 +1,209 @@
...
+            title: 'Відмінити зміни',
+          },
+        ]}
+      />
+      <VideoWrapper duration="38m 2s">
+        <iframe
+          src="https://www.youtube.com/embed/HJ7NW6zatXo"
+          allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture"
+          allowFullScreen
+        ></iframe>
```
Comment (liza-veis) [inline src/components/lectures/how-to-work-with-git/ua/index.tsx]:
```suggestion
        />
```

File: src/enums/code-language/code-language.ts [hunk 16]
```diff
@@ -0,0 +1,13 @@
+enum CodeLanguage {
+  JS = 'javascript',
+  JSON = 'json',
+  TEXT = 'text',
+  CSHARP = 'csharp',
+  JAVA = 'java',
```
Comment (liza-veis) [inline src/enums/code-language/code-language.ts]:
```suggestion
```

File: src/layouts/layout/layout.tsx [hunk 17]
```diff
@@ -0,0 +1,15 @@
+import React, { PropsWithChildren } from 'react';
+import { Footer, Header } from '~/partials';
+
+import * as styles from './layout.module.scss';
+
+const Layout: React.FC<PropsWithChildren> = ({ children }) => (
```
Comment (liza-veis) [inline src/layouts/layout/layout.tsx]: define props via children

File: src/pages/index.tsx [hunk 18]
```diff
@@ -0,0 +1,9 @@
+import * as React from 'react';
+import { PageProps } from 'gatsby';
+import { MainPage } from '~/components/main-page/main-page';
+import { SEO } from '~/partials';
+
+const Homepage: React.FC<PageProps> = () => <MainPage />;
```
Comment (liza-veis) [inline src/pages/index.tsx]:
```suggestion
const MainPage: React.FC<PageProps> = () => <MainPage />;
```

File: src/partials/seo/seo.tsx [hunk 19]
```diff
@@ -1,12 +1,12 @@
...
 import { siteMetadata } from './meta/default-meta';
 
-type SEOProps = {
-  pageContext?: PageContext;
+type Props = {
+  pageContext?: SeoProps;
 };
 
-const SEO: React.FC<React.PropsWithChildren<SEOProps>> = ({
+const SEO: React.FC<React.PropsWithChildren<Props>> = ({
```
Comment (liza-veis) [inline src/partials/seo/seo.tsx]: define props via children

**Review by liza-veis — CHANGES_REQUESTED** [review]

File: src/components/common/copy/copy.tsx (hunk 8 above)
Comment (kulak91) [reply src/components/common/copy/copy.tsx:18]: I didn't want to leave any comment, but this is too much. It gives better UX.

File: src/components/common/level/level.tsx (hunk 10 above)
Comment (kulak91) [reply src/components/common/level/level.tsx:12]: Scroll view

File: src/components/common/panels/panels.module.scss (hunk 11 above)
Comment (kulak91) [reply src/components/common/panels/panels.module.scss]: They are fixed. DRY ? Ok. No

File: src/components/common/panels/panels.tsx (hunk 12 above)
Comment (kulak91) [reply src/components/common/panels/panels.tsx:21]: Any reasons? It's convenient since there are different columns grid on the page.
Component should be reusable, right?
Following this logic what is better: specifying numbers of columns through the props or adding 
unnecessary  classname somewhere else?
This component is based on the grid layout.

File: src/components/common/panels/panels.tsx (hunk 14 above)
Comment (kulak91) [reply src/components/common/panels/panels.tsx:18]: Yes we do. Same as columns

File: src/pages/index.tsx (hunk 18 above)
Comment (kulak91) [reply src/pages/index.tsx]: Const MainPage which exports MainPage?

File: src/partials/seo/seo.tsx (hunk 19 above)
Comment (kulak91) [reply src/partials/seo/seo.tsx]: Reasons?

File: src/components/common/copy/copy.tsx (hunk 8 above)
Comment (liza-veis) [reply src/components/common/copy/copy.tsx:18]: oh didn't notice that it's for ui, let's leave it and just move 1500 to constants at the top of file

File: src/components/common/level/level.tsx (hunk 10 above)
Comment (liza-veis) [reply src/components/common/level/level.tsx:12]: ok, let's leave it for noe

File: src/components/common/panels/panels.tsx (hunk 12 above)
Comment (liza-veis) [reply src/components/common/panels/panels.tsx:21]: Don't like this approach, will need to think how to get rid of it, let's leave for now, but we will definitely remove this in the future

File: src/pages/index.tsx (hunk 18 above)
Comment (liza-veis) [reply src/pages/index.tsx]: then name it as just Page, we don't want to have two names for same page

File: src/partials/seo/seo.tsx (hunk 19 above)
Comment (liza-veis) [reply src/partials/seo/seo.tsx]: 1. more flexible, you can define that it should be an array with two nodes etc
2. we use this approach elsewhere in academy projects
3. looks more obvious

File: src/components/common/panels/panels.module.scss (hunk 11 above)
Comment (liza-veis) [reply src/components/common/panels/panels.module.scss]: Don't want to use scss features like this, it's ok when in some cases you repeat code, you can't get rid of repeating at all

**Review by liza-veis — APPROVED** [review]
